SEO - May (#1464)
moonbeam-foundation/moonbeam-docs: builders/integrations/indexers/subsquid.md

## Introduction {: #introduction }

[SQD (formerly Subsquid)](https://sqd.dev/){target=\_blank} is a data network that allows rapid and cost-efficient retrieval of blockchain data from 200+ chains using the decentralized data lake and open-source SDK from SQD. In very simple terms, you can think of SQD as an ETL (extract, transform, and load) tool with a GraphQL server included. It enables comprehensive filtering, pagination, and even full-text search capabilities.

> SQD has native support for both Ethereum Virtual Machine (EVM) and Substrate data. Since Moonbeam is a Substrate-based smart contract platform that is EVM-compatible, you can use SQD to index both EVM and Substrate-based data. The Squid SDK currently offers specialized processors for EVM (`EvmBatchProcessor`) and Substrate (`SubstrateBatchProcessor`) networks. If you exclusively want to index EVM data, use the EVM processor. If you need Substrate events or calls, or Frontier EVM data alongside Substrate context, use the Substrate processor and the Frontier EVM utilities.

Reviewer: ```suggestion
SQD has native support for both Ethereum Virtual Machine (EVM) and Substrate data. Since Moonbeam is a Substrate-based smart contract platform that is EVM-compatible, you can use SQD to index both EVM and Substrate-based data. If you exclusively want to index EVM data, use the EVM processor. If you need Substrate events or calls, or Frontier EVM data alongside Substrate context, use the Substrate processor and the Frontier EVM utilities.
```

I think we should drop the technical implementation details like `EvmBatchProcessor` since it's the introduction, should just be high level. Some of it just feels repetitive too so recommended removing some of it.

This quick-start guide shows you how to create Substrate and EVM Squid SDK projects with SQD and configure them to index data on Moonbeam.

--8<-- 'text/_disclaimers/third-party-content-intro.md'

## Checking Prerequisites {: #checking-prerequisites }

Before you get started with SQD, install the following:

- [Node.js](https://nodejs.org/en/download){target=\_blank} version 18 or newer

> - [Git](https://git-scm.com/downloads){target=\_blank}

Reviewer: We don't typically add git as a prereq. Doesn't look like this guide specifically use git either (only took quick look), so I think this can be removed

- [Docker](https://docs.docker.com/get-started/get-docker/){target=\_blank}
- [Squid CLI](https://docs.sqd.dev/en/cloud/reference/cli/installation){target=\_blank}

!!! note

>     The commands in this guide use `npm`, matching the current Squid SDK templates and quick-start docs from SQD.

Reviewer: This note seems unnecessary if we're just using it to callout what is being used. If yarn still doesn't work, we should call that out though

Reviewer: Also putting npm and yarn in backticks is not necessary anymore

    === "Moonbeam"

        ```ts
        const processor = new SubstrateBatchProcessor()
          .setGateway('https://v2.archive.subsquid.io/network/moonbeam-substrate')
          .setRpcEndpoint('{{ networks.moonbeam.rpc_url }}')
        ```

    === "Moonriver"

        ```ts
        const processor = new SubstrateBatchProcessor()
          .setGateway('https://v2.archive.subsquid.io/network/moonriver-substrate')
          .setRpcEndpoint('{{ networks.moonriver.rpc_url }}')
        ```

    === "Moonbase Alpha"

        ```ts
        const processor = new SubstrateBatchProcessor()
          .setGateway('https://v2.archive.subsquid.io/network/moonbase-substrate')
          .setRpcEndpoint('{{ networks.moonbase.rpc_url }}')
        ```

    !!! note
        For the Moonbeam or Moonriver RPC endpoint passed to `setRpcEndpoint()`, use your own endpoint from one of the supported [Endpoint Providers](/builders/get-started/endpoints/){target=\_blank}. This RPC endpoint is separate from the SQD gateway configured with `setGateway()`.

> 5. Make one more quick change to the template. The SQD Substrate template processes Substrate account types by default, but Moonbeam uses Ethereum-style accounts. The `getTransferEvents` function in the `src/main.ts` file iterates through the events ingested by `processor.ts` and stores the relevant `transfer` events in the database. In the `getTransferEvents` function, remove the ss58 encoding of the `from` and `to` fields. An unmodified Substrate template ss58 encodes the `from` and `to` fields as shown:

Reviewer: Because the first sentence is now action oriented instead of just saying "hey, there's one more change", and we're not stating the change, it seems a little more unnecessary than it was before haha

```suggestion
5. The SQD Substrate template uses Substrate-style accounts by default, but Moonbeam uses Ethereum-style accounts, so you'll need to remove the ss58 encoding from the account fields. The `getTransferEvents` function in `src/main.ts` is responsible for iterating through the events ingested by `processor.ts` and saving the relevant transfer events to the database — this is where you'll make that change. Remove the ss58 encoding from the `from` and `to` fields. The default, unmodified Substrate template encodes these fields as shown:
```

    ```ts
    from: ss58.codec('kusama').encode(rec.from),
    to: ss58.codec('kusama').encode(rec.to),
    ```

    After you remove the ss58 encoding, use the following lines:

    ```ts
    from: rec.from, 
    to: rec.to, 
    ```

> After you configure your SQD project to index Substrate data on Moonbeam, update the `schema.graphql`, `typegen.json`, `src/main.ts`, and `src/processor.ts` files to index the data you need for your project. Next, take the steps in the [Run Your Indexer](#run-your-indexer) section to run your indexer and query your Squid.

Reviewer: The phrasing "After you configure your SQD project..." implies the configuration is something the user still needs to do, but they've just done it. So that sort of tone is lost in this change.

```suggestion
Now that your SQD project is configured to index Substrate data on Moonbeam, update the `schema.graphql`, `typegen.json`, `src/main.ts`, and `src/processor.ts` files to index the data you need for your project. Next, take the steps in the [Run Your Indexer](#run-your-indexer) section to run your indexer and query your Squid.
```

## Index Ethereum Data on Moonbeam {: #index-ethereum-contracts }

To index EVM data on Moonbeam, create a SQD project and configure it for Moonbeam by taking the following steps:

> 1. Create a SQD project for EVM data by using the generic EVM template or the ABI template. The EVM template is a minimal Squid SDK project for EVM data. Use the ABI template with [Squid generation tools](https://docs.sqd.dev/en/sdk/squid-sdk/resources/tools/squid-gen){target=\_blank} to generate a squid from contract ABIs:

Reviewer: ```suggestion
1. Create a SQD project for EVM data by using the generic EVM or ABI template, which works with [Squid generation tools](https://docs.sqd.dev/en/sdk/squid-sdk/resources/tools/squid-gen){target=\_blank} to generate a squid from contract ABIs:
```

The closing sentence was only focusing on the ABI template, which makes it seem like that's the only option. Both templates should be represented equally.

        sqd init INSERT_SQUID_NAME --template abi
        ```

>     For more information on getting started with Squid SDK templates, check out the [Development flow](https://docs.sqd.dev/en/sdk/squid-sdk/how-to-start/squid-development){target=\_blank}.

Reviewer: ```suggestion
    For more information on getting started with Squid SDK templates, check out the [Development flow](https://docs.sqd.dev/en/sdk/squid-sdk/how-to-start/squid-development){target=\_blank} guide.
```

2. Navigate into the root directory of your Squid project and install dependencies by running:

    ```bash
    npm i
    ```

3. If you use the ABI template, generate the squid code for the contract and events or functions you want to index:

>     ```bash

Reviewer: So to follow up on my previous comment, this command generates the squidgen.yaml? or what?

what is the "squid code" that is being generated here?

        sqd init INSERT_SQUID_NAME --template abi
        ```

>     For more information on getting started with Squid SDK templates, check out the [Development flow](https://docs.sqd.dev/en/sdk/squid-sdk/how-to-start/squid-development){target=\_blank} and [Squid SDK Quickstart](https://docs.sqd.dev/en/sdk/squid-sdk/quickstart){target=\_blank} guides on the SQD documentation site.

Reviewer: Are these the best links? Ideally we could continue to point to one method or the other

2. Navigate into the root directory of your Squid project and install dependencies by running:

    ```bash
    npm ci
    ```

> 3. If you use the ABI template, configure the contracts and events you want to index in `squidgen.yaml` and generate the squid code:

Reviewer: How are they supposed to configure the `squidgen.yaml`?

    ```bash
    npx squid-gen config squidgen.yaml
    ```

>     The `squidgen.yaml` configuration includes the `archive` field for the SQD Network gateway, the target data store, and the contracts, ABIs, events, and functions to index. For the `archive` value, use the corresponding Moonbeam EVM gateway from the next step.

Reviewer: ```suggestion
    The `squidgen.yaml` configuration includes the `archive` field for the SQD Network gateway, the target data store, and the contracts, ABIs, events, and functions to index. For the `archive` value, use the corresponding Moonbeam EVM gateway used in the next step.
```

> 4. Modify the `src/processor.ts` file, which is where Squids instantiate the processor, configure it, and attach handler functions. The processor can fetch historical on-chain data from a SQD Network gateway. Configure your processor to pull data from the [EVM dataset](https://docs.sqd.dev/en/data/evm){target=\_blank} that corresponds to the network you are indexing data on:

Reviewer: ```suggestion
4. Modify the `src/processor.ts` file, which is where Squids instantiate the processor, configure it, and attach handler functions. The processor fetches historical on-chain data from a SQD Network gateway. Configure your processor to pull data from the [EVM dataset](https://docs.sqd.dev/en/data/evm){target=\_blank} that corresponds to the network you are indexing data on:
```

    === "Moonbeam"

        ```ts
        const processor = new EvmBatchProcessor()
          .setGateway('https://v2.archive.subsquid.io/network/moonbeam-mainnet')
        ```

    === "Moonriver"

        ```ts
        const processor = new EvmBatchProcessor()
          .setGateway('https://v2.archive.subsquid.io/network/moonriver-mainnet')
        ```

    === "Moonbase Alpha"

        ```ts
        const processor = new EvmBatchProcessor()
          .setGateway('https://v2.archive.subsquid.io/network/moonbase-testnet')
        ```

> After you configure your SQD project to index EVM data on Moonbeam, update the `schema.graphql`, `src/main.ts`, and `src/processor.ts` files to index the data you need for your project. Continue with the steps in the following section to run your indexer and query your Squid.

Reviewer: ```suggestion
Now that you've configured your SQD project to index EVM data on Moonbeam, update the `schema.graphql`, `src/main.ts`, and `src/processor.ts` files to index the data you need for your project. Continue with the steps in the following section to run your indexer and query your Squid.
```

## Run Your Indexer {: #run-your-indexer }

> These steps apply to both Substrate and EVM indexers. After you configure your SQD indexer, run it by taking the following steps:

Reviewer: Have you gone through all of these steps?

Reviewer: yes? no? maybe so? haha

Author: Hey, yes, I did, but I miss the abi template. I had to make some changes to the guide.

!!! note

>     The `setGateway()` examples in this guide use legacy `https://v2.archive.subsquid.io` gateways. SQD requires an API key for self-hosted Squid SDK projects that use these legacy gateways. SQD Cloud squids that rely on legacy gateways are not affected. See the [Accessing SQD data with API keys](https://docs.sqd.dev/en/data/api-keys){target=\_blank} guide for more information.

Reviewer: By not affected you mean they don't need to use an API key? I guess I'm not sure what you mean ehre

1. Build the project by running:

    ```bash

>     npm run build

Reviewer: There isn't a sqd command? Like what about this (just found and copied from their docs):

```
sqd build             Build the squid project
```

    ```

2. Launch Postgres by running:

    ```bash

>     docker compose up -d

Reviewer: ```
sqd up                  Start a local PG database
```

    ```

3. Apply database migrations:

    ```bash

>     npx squid-typeorm-migration apply

Reviewer: ```
sqd migration:apply             apply pending migrations
```

    ```

    If you changed the schema or generated an ABI-based squid, generate migrations before applying them:

    ```bash
    npx squid-typeorm-migration generate

>     npx squid-typeorm-migration apply

Reviewer: ```
sqd migration:apply             apply pending migrations
sqd migration:generate          generate the migration for the schema defined in schema.graphql
sqd migration:clean             clean the db/migrations folder
```

    ```

4. Inspect and run the processor:

    ```bash
    node -r dotenv/config lib/main.js
    ```

5. Open a separate terminal window in the same directory, then start the GraphQL server:

    ```bash

>     npx squid-graphql-server

Reviewer: ```
sqd run [PATH]          Run all the services defined in squid.yaml locally
```
